### ISSUE TITLE
Add a mini-engine to NUnit

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#571]
The notion of a mini-engine has come up in several separate contexts, in particular in relation to the Visual Studio adapter and the compact framework build.

Our current engine does a lot of things and not all of them make sense when used by a special-purpose runner or in a device. In addition, some features are not used when the engine is used in an nunit-agent process.

I'm labelling this an idea with status:design since we need to define what we mean by a mini-engine before we can imlement it - even before we can say whether we need it.

I'm assigning this to myself for the design phase and I'll add specific design issues as individual comments. Others should do the same to comment on or add to the design. The milestone assigned is intended to represent completion of design and making a decision. Implementation can take longer.

We haven't tried this mechanism for iterating to a design before. Let's give it a chance and see what happens.

**Comment by CharliePoole** [comment]
What the VS adapter would need from a mini-engine:

* Identify whether an assembly is an NUnit 3 test assembly.
* Have a runner that can load and run a single assembly in a test domain.

Would not need:

* Mono-addins
* V2 driver
* Runners for execution in another process.
* Ability to control the framework used for execution
* Creation of XML output

**Comment by CharliePoole** [comment]
**Primary and Secondary Engines**

So we can talk about this more readily, I'm inventing some terms.

  * The **Primary Engine** is the engine that is acquired and used by a runner, such as nunit-console.
  * A **Secondary Engine** is one that is communicated with and controlled by a **Primary Engine**.

There are certain features that are not needed in a secondary engine, simply because they are  taken care of by the primary engine. For example, a secondary engine does not need to create result reports because the xml result is passed back to the primary, which has the job of creating the reportss. Hence, all code for report generation can be removed from an engine if it is solely used as a secondary engine.

We can consider four possible combinations of the engine type and it's role:

  * Full Engine as primary: This is the normal case of the engine under nunit-console

  * Full Engine as secondary: This is how we use the engine under nunit-agent. Some features are suppressed by the way the engine is initialized, but the full engine is still loaded in the process.

  * Mini-Engine as primary: We would want this if any of our device runners were to use the engine. It's also how the VS adapter could work with a mini-engine.

  * Mini-Engine as secondary: We could use this configuration if we introduced a communications protocol to allow the desktop primary to connect with a secondary engine on a device.

**Comment by CharliePoole** [comment]
I created a mini-engine build, leaving out all addins, some services and out of process runners. Here's a tally of assembly sizes.

Full Engine plus standard addins
  nunit.engine              79K
  nunit.engine.api        14K
  mono.addins           319K
  nunit.v2.driver           17K
  nunit.core                136K
  nunit.core.interfaces  56K
  nunit-project-loader   11K
  vs-project-loader       14K
  nunit-v2-result-writer 16K
Total                          662K

Mini-Engine
  nunit.mini-engine       53K
Total                            53K

It's clear that the biggest reduction in size by far is from removing addins. Of course, you lose that functionality when you do it.

**Comment by oznetmaster** [comment]
Why would you remove out of process runners?  I am confused.

**Comment by CharliePoole** [comment]
Because it's not needed for some of the applications I was thinking of, which already set up the process in which the tests are to be run.

But anyway, it's not a significant factor in the size of the assembly as you see from the numbers.

**Comment by oznetmaster** [comment]
I would like to see what you did, since my “min-engine” is about 135K under CF.  It also has no add-ins, and no remoting.

**Comment by CharliePoole** [comment]
I pushed it out to branch mini-engine-experiment. It's now somewhat behind master.

**Comment by CharliePoole** [comment]
@nunit/core-team Here's another of these "ideas" that tends to just sit around for a while. As shown in the branch, it's feasible to have a mini-engine with limited functions for use in special contexts like the vs adapter and possibly within certain runners for devices. I think it could be made to compile as a portable library as well.

Let's try to decide if we want such a thing and then either schedule this to a milestone or drop it.

**Comment by oznetmaster** [comment]
I have not been able to make a reasonable comment on your mini-engine, since it is quite different than mine, and I am trying to reconcile the decisions that each of us made as to what to leave in and what to take out (mine is about twice the size of yours, and has more functionality.)

 

There may even be a case for a min-engine, a midi-engine, and a maxi-engine.  Same source, just conditionals.

 

I will comment in the near future.

**Comment by CharliePoole** [comment]
Hi Neil,

I'm not looking for code comments. I'd just like us to decide whether this
is something we want to develop. I don't think we should spend any more
time on it unless we do.

Charlie

**Comment by oznetmaster** [comment]
I was not talking about code comments.  I am trying to determine functionality for each version.  What it would be used for and how it would be used.

**Comment by CharliePoole** [comment]
The question is whether we should have anything at all, not whether some
specific functionality should be in it. I only pushed a branch with my
experiment at your request and I think you're the only one looking at the
details in this way.

I am imagining we will have decided to drop the idea before we hear from
you but I hope not.

**Comment by oznetmaster** [comment]
You give the impression that there is some deadline for making a decision.  As you mentioned, your take on a mini-engine was an experiment, as is mine.  I am continuing to work with it.  I did not realize that a decision on to have or not to have was imminent :(

**Comment by CharliePoole** [comment]
I had hoped my comment about "ideas" here and on recent emails would give you an idea of the sense of urgency I feel. Of course, any time spent on experiments has to be limited for a project to move ahead, but this one is particularly problematic:

1. We're not experimenting about _how_ to do this but _whether_ it's a useful idea.

2. While we wait, other tasks that might use such a mini-engine are being completed without it, making the decision eventually moot. The VS adapter has already been released as a CTP using more or less adhoc code to invoke the framework. The IoS and Android runners are being developed without without use of an engine.

This is a fairly central decision to the project. It basically decides whether the runner/engine/framework architecture is going to apply to the desktop platforms only or to all platforms. So, yes, I do think it's fairly urgent.

**Comment by oznetmaster** [comment]
It is my personal feeling that some form of runner/engine/framework architecture should apply to all platforms.  I could have written an ad hoc interface for the CF version, but have chosen to try to develop a model for a more general structure.  The minute we start developing ad hoc interfaces for each different platform, we lose the consistency which has seemingly been one of the driving foundations of NUnit.

**Comment by rprouse** [comment]
I am not sure how useful a mini-engine will be for the Xamarin platforms. I just create an `NUnitTestAssemblyRunner`, execute it and parse the `ITestResult` that is returned. The only thing that would have been useful is a class to summarize the `ITestResult`. I based my [ResultSummary.cs](https://github.com/nunit/nunit.runners/blob/master/src/runner/nunit.runner/ViewModel/ResultSummary.cs) off the one in nunitlite.

For these other platforms, command line options, spawning test processes, and everything else an engine provides is not useful. I found nearly everything I need right in the framework with a fairly clean API.

I also don't see much need for it on the desktop. We already have a functional engine there.

Maybe this mini-engine is just needed for CF?

**Comment by CharliePoole** [comment]
So, the runners for Xamarin are more nunitlite-like - they reference a particular framework version. That means, of course, we will have to release a new runner with each new framework release. Is that ideal? The main reason a mini-engine might be useful for the VS adapter is to keep the release cycles separate.

**Comment by CharliePoole** [comment]
Maybe we need some sort of matrix of engine capabilities versus potential applications to help guide this.

**Comment by rprouse** [comment]
@CharliePoole, that is a good point about versioning. Would it work if I set Specific Version on the reference to false? I was planning on doing that and then in the NuGet package specifying a range of versions for the dependency.

**Comment by CharliePoole** [comment]
You'll have to try it. I suspect that setting applies to compile-time only. One main purpose for a small version of the engine would be to provide this insulation. Essentially, that's what the driver does in the engine. Maybe I'm talking about a micro-engine here. :-)

**Comment by oznetmaster** [comment]
That is exactly what I was thinking.  This discussion has been a bit too abstract.

 

What does each platform need from its local “engine”?

**Comment by oznetmaster** [comment]
Maybe, for many of the smaller platforms (CF, for instance), what is needed is not really an engine, but a “driver”.  Or are they the same?  Micro, mini, midi, full?  Too many engines for me to contemplate :(

**Comment by CharliePoole** [comment]
I created a table of engine features as used by various runners.

@oznetmaster Please review the CF column and adjust as needed.

@rprouse  Please add a column for Xamarin (or two if they differ)

Looking at it, I so far only see the need for what we might call a "micro-engine" The experimental mini-engine I created has way too much in it for use with the VS adapter. Basically, all that's needed is a simple decoupling of the runner from the framework without the need to use reflection.
Its seems like the Xamarin runners would not need much more.

Such an engine might also be attractive for use by third-party runners.

I have a sense that @oznetmaster may want something more than this for CF, however. ;-)

**Comment by CharliePoole** [comment]
@oznetmaster In our architecture, the Engine provides Runners and Runners use Drivers to talk to the framework. My adhoc code for the NUnit 3 Test Adapter uses the driver directly. But it takes a bit of surrounding code to do that. The smallest engine I can imagine would simply bundle that code in an easy to use package and put an ITestEngine interface on it.

For CF, I think it depends on how much we want to provide as opposed to enabling others to do.

**Comment by oznetmaster** [comment]
Where is this table?

**Comment by oznetmaster** [comment]
What I need is the ability to use nunit-console, and whatever GUI is developed, unchanged, communicating with a CF NUnit running on a device.

**Comment by CharliePoole** [comment]
Well, that's definitely our main thrust. But you have been pushing for something that runs on CF - or so I understood. Do you no longer want to have a true test runner that runs on CF? That was a big part of my reason for proposing a mini-engine.

**Comment by oznetmaster** [comment]
I need both.  They are not mutually exclusive.  And I assume that the mini/micro engine would be required in either case?

**Comment by oznetmaster** [comment]
To clarify, there needs to be a console runner that runs natively on CF, and a engine?? based runner/driver/engine which connects with the existing PC based infrastructure.

**Comment by oznetmaster** [comment]
I thought that this was what we had started talking about when we were discussing remote testing (a development pc running tests on a remote test pc).  It is really the same thing, but with a more limited capability set for the CF remote device.

**Comment by CharliePoole** [comment]
No, they aren't mutually exclusive. At the moment, I was focused on running totally on the platform, but I imagine the same features are needed in either case. Wouldn't you agree?

So, if you need a console runner on CF, let's imagine it uses an engine. What features would that engine have to include?

If we can get that out of the way, we can focus on remote testing, which is what I _really_ want to focus on. Remote testing may use an engine on the device. I see two ways to go: engine-based and driver-based. I would want to do some spikes to decide which works best.

Engine-based: The desktop engine talks to an engine on the device, which in turn uses a driver to call the framework.

Driver-based: A driver in the desktop engine talks directly to the framework on the device.

Frankly, remote testing is much more interesting to me than running directly on CF. That's why I'd like to nail down how far we need to go in running on CF and then knock out the code. IMO it really should not be terribly hard.

**Comment by oznetmaster** [comment]
I am happy with running on CF using the (slightly) modified version of AutoRun that I am currently using (the one that allows the name of the test assembly to be entered on the command line).  It seems adequate.

I am not sure what more is really needed.  Many things have been mentioned, but they may be unnecessary.  Even more so if there is a true remote testing option.

The only clear reason for local testing on CF devices (once remote testing is available) is in a disconnected environment, where the tests may be loaded from a USB key or some other media, or for preloaded tests that are distributed with the device, or the software being tested.

I am currently focused on remote testing, since the local testing solution I have is more then adequate.

**Comment by rprouse** [comment]
I have updated the table with the Xamarin information. You were right @CharliePoole, since the framework is signed (damn you Microsoft), I need a micro-engine to handle loading of different versions of the framework. Either that, or we don't sign the portable framework, but I don't think that is feasible.

**Comment by oznetmaster** [comment]
What table.  I cannot find it to update it :(

**Comment by rprouse** [comment]
@oznetmaster, the table is part of the issue we are discussing. You probably just missed it in the emails that are being sent out. The table is in the wiki at https://github.com/nunit/dev/wiki/Mini-Engine

**Comment by CharliePoole** [comment]
@rprouse  Your runners need even less than the VS adapter - leaving aside remote communication. I need to set up a test domain because I don't control the ApplicationBase under which VS calls the adapter. It sounds like the Xamarin runners function very much like NUnitLite, with the addition of a layer that eliminates the need for a reference.

Now that @oznetmaster has got me thinking more about it, using a mini-engine as a communications partner for the desktop engine makes perfect sense. I had been thinking about using a special driver that knew how to communicate directly to the framework. I'll need to try out both options, but it seems like a device-based engine would iron out the differences between the various devices rather neatly.

@oznetmaster Looking at the entries you added to the table, I'm wondering about reading nunit or VS projects on the CF device. I'd expect them to be resolved to a list of assemblies by the time the device was asked to execute them. But you might have a use case I don't know about.

Regarding multiple processes... can you launch processes programatically in CF?

**Comment by oznetmaster** [comment]
CF supports the Process class in the System.Diagnostics namespace, with some limitations, but those limitations do not seem to impact NUnit.  It also provides limited AppDomain support, perhaps not enough.  All you can do in an AppDomain is to load an assemly, and run it at its entry point, getting back an integer result. No SetupInformation.  CF does support multiple simultaneous AppDomain.  It may not be enough to allow AppDomain to be used at all in CF, but I am looking at it closer.

**Comment by oznetmaster** [comment]
I am sorry, I misunderstood.  I want NUnit projects and VS projects to be able to be used with CF NUnit tests, but not locally.  Only for remote testing.

**Comment by oznetmaster** [comment]
I have been focusing on using a mini-engine as the communications partner.  The problem is that the current engine/mini-engine does not adequately factor out the communications methodology between the two engines to allow it to be replaced easily.  Remoting is fairly embedded.  So that seems to be the first step.

**Comment by CharliePoole** [comment]
@oznetmaster I updated the netcf column to remove the projects. I removed the ?? next to Multiple (different) frameworks. This refers to being able to run tests using other frameworks than NUnit 3.0. 

I also replaced the No entries with blanks - I hope the visual volume of Yes entries will show how "heavy" each use case is.

It seems to break down into needing a _very_ lightweight engine for the adapter and xamarin, with a slightly heavier one for CF. Both of those might eventually add some communications capability, but I agree with @rprouse that has to be post-3.0 if we want to get to a final release any time soon.

[BTW, once we get a final 3.0, I hope we can move to much more frequent releases.]
I added a column for nunit-agent. The features it uses match closely to what CF needs. We currently use the same engine under agent as under the console and just don't make use of those features. Some of them are actually locked out by the way we initialize the engine. I didn't want to create a separate engine build just for the agent, but if we had one that met it's needs, we might use it. Note, however that support for V2 tests currently requires Mono.Addins. We could change this, of course.

[A bigger question is whether we are getting sufficient benefit out of mono addins to justify using it. We can save that for another time.]

**Comment by CharliePoole** [comment]
I rebased my experimental branch to bring it up to date: https://github.com/nunit/nunit/tree/mini-engine-experiment.

I'm considering two alternative changes:

1. Using feature defines instead of just MINI_ENGINE. That would be things like PROJECT_SERVICE, SETTINGS_SERVICE, etc. The only drawback is that there could potentially be a lot of them.

2. Creating a Core Engine, with the minimal set of features needed for the adapter and the Xamarin runners. It could share source with the full engine or the full engine could actually inherit from it.

**Comment by oznetmaster** [comment]
I like both options.  I do not think they are mutually exclusive?  One feature define would then be CORE_ENGINE.

**Comment by CharliePoole** [comment]
@oznetmaster No they aren't. CORE_ENGINE could represent a specific set of features, just as MINI_ENGINE does in the current branch. However, I was thinking of abandoning MINI_ENGINE and _only_ specifying actual features - similar to the lines in the wiki table.

For option 2, I imagined an nunit.engine.core assembly, containing a CoreEngine class, which could be used as is or inherited from. CoreEngine would be the baseline - the engine with the minimal set of features below which we don't go. It would be even smaller than what I already have. 

If we went that way, I would imagine that the compiler defines would only go into more full-featured, derived engines.

**Comment by CharliePoole** [comment]
I'm starting work on a CoreEngine and will push a new experimental branch for it.

**Comment by rprouse** [comment]
@CharliePoole, I finally got a chance to go through the code for your mini-engine experiment tonight. We are still a long way off from having anything portable. Classes like AppDomain, File, Path, etc are not portable. Since variations are often available on various platforms, I have seen dependency injection used to create minimal interfaces for these things. For example, a minimal IAppDomain that is implemented and injected on a platform by platform basis.

The problem is that Android and iOS are such locked down environments. It is like running Silverlight in the browser. Windows store apps and the new Win10 universal apps are similar.

If we are doing this for all to support all of those platforms, then I think that a mini-engine needs to nearly start from scratch, pass in assemblies to be tested and return test results using base types like string and int. 

Of course, the problem with that is how do we break the dependencies on a specific version of the framework with this model? Do we abandon the idea of one portable mini engine and create platform specific versions? Is there another option that is staring me in the face and I am not seeing?

The Xamarin runners are very nunitlite like, so we could release them in step with the framework, but I don't think that is the way we want to go with all platforms.

Sorry, a bit rambling and without solutions, I just wanted to document my thoughts in this issue for when we come back to it.

**Comment by CharliePoole** [comment]
I had pretty much set the mini-engine aside in favor of the core engine, with fewer features. Neil still had some interest in it for CF and I believe has created a private build for his purposes.

It would help if we could characterize the key issues on these platforms. What can you / can't you do? On Silverlight, I believe it's not possible to specify a file name from which to load, but you can load using an assembly name. It's academic however, as we have no command-line or gui there in which to specify what is to be loaded. 

Maybe what we want to look at is an nunitlite-like model without the use of reflection, using the engine only to run tests in the already-loaded assembly without the need for a reference to the framework. This would be pretty easy to do starting from the current core engine and removing features. I guess all services could go and there would be only one driver to select.

So, in spite of your frustration, I think you are circling around a possible solution... unless it's the drain you're circling. :-)

**Comment by CharliePoole** [comment]
I think this is pretty much all resolved. I need to go through and see if there are any loose ends before closing it.

**Comment by cwensley** [comment]
This has been resolved for me, the mini engine works wonders!  (I'm using it [here](https://github.com/picoe/Eto/blob/develop/Source/Eto.Test/Eto.Test/Sections/UnitTestSection.cs)) in a PCL assembly, and runs on all desktop and mobile platforms on mono, xamarin, and .NET.

**Comment by CharliePoole** [comment]
That looks cool! Glad it's working for you.

In fact you appear to be using the framework directly, rather than what we call the engine. That's how we run tests on multiple devices as well.

**Comment by CharliePoole** [comment]
@nunit/core-team I'm about to close this.

I think we got a lot of value out of the  discussion and no longer plan to have a separate "mini" engine build. Increasing the degree to which functionality is provided by extensions will give us the ability to eliminate that functionality through configuration rather than by having a special build.

I'll give you a bit of time to comment back and then close this as wontfix.
